### PR TITLE
Targeting API 34, Gradle 8.4 and fixes the main issues and deprecations

### PR DESCRIPTION
**Pull request by stefan-tudor**
**Migrations:**
- from Groovy Gradle scripts to Kotlin scripts
- a separate project for convention plugins and version catalogs

**Adds:**
- Detekt
- Ktlint
- Dokka
- Kotlin binary compatibility validator
- Firebase Crashlytics

Targeting API 34, Gradle 8.4 and fixes the main issues and deprecations

### CODE REVIEW

**Review by eugenezhernakov — APPROVED** [review]

File: build-logic/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+which is most useful for configuring Android components (libraries vs applications) with shared
+code.
+
+These plugins are *additive* and *composable*, and try to only accomplish a single responsibility.
+Modules can then pick and choose the configurations they need.
+If there is one-off logic for a module without shared code, it's preferable to define that directly
+in the module's `build.gradle`, as opposed to creating a convention plugin with module-specific
+setup.
+
+Current list of convention plugins:
```
Comment (RafalOzogJudopay) [inline build-logic/README.md:30]: Seems like line 30 maybe can be removed (as no listed items present).

File: judokit-android/api/judokit-android.api [hunk 2]
```diff
@@ -0,0 +1,4989 @@
...
+	public final fun unRegisterResultListener (Lcom/judopay/judokit/android/service/CardTransactionManagerResultListener;Ljava/lang/String;)V
+	public static synthetic fun unRegisterResultListener$default (Lcom/judopay/judokit/android/service/CardTransactionManager;Lcom/judopay/judokit/android/service/CardTransactionManagerResultListener;Ljava/lang/String;ILjava/lang/Object;)V
+	public fun updateActivity (Landroidx/fragment/app/FragmentActivity;)V
+}
+
+public final class com/judopay/judokit/android/service/CardTransactionManager$Companion : com/judopay/judokit/android/service/SingletonHolder {
+}
+
+public abstract interface class com/judopay/judokit/android/service/CardTransactionManagerResultListener {
+	public abstract fun onCardTransactionResul
```
Comment (RafalOzogJudopay) [inline judokit-android/api/judokit-android.api:4989]: I am actually wondering what is this file tbh (I understand it's our api interface, but wondering what is the use case for this), any hint would be appreciated ;)

File: judokit-android/src/main/java/com/judopay/judokit/android/api/AppMetaDataProvider.kt [hunk 3]
```diff
@@ -17,14 +16,15 @@ class AppMetaDataProvider(context: Context, subProductInfo: SubProductInfo) {
...
-        } catch (e: PackageManager.NameNotFoundException) {
-            e.printStackTrace()
-            ""
-        }
+        get() =
+            try {
+                val packageInfo = applicationContext.packageManager.getPackageInfo(applicationContext.packageName, 0)
+                packageInfo.versionName
+            } catch (e: PackageManager.NameNotFoundException) {
+                ""
```
Comment (RafalOzogJudopay) [inline judokit-android/src/main/java/com/judopay/judokit/android/api/AppMetaDataProvider.kt:26]: Just to confirm - we don't expect to handle any catch case here (as stacktrace printing has been removed), is that correct?

File: judokit-android/src/main/java/com/judopay/judokit/android/model/RecommendationConfiguration.kt [hunk 4]
```diff
@@ -43,29 +40,29 @@ class RecommendationConfiguration internal constructor(
...
-                "The RSAPublicKey field in the ravelin recommendation configuration is required."
-            )
+            val myRsaKey =
+                requireNotNullOrEmpty(
+                    rsaPublicKey,
+                    "rsaPublicKey",
+                    "The RSAPublicKey field in the ravelin recommendation configuration is required.",
+                )
 
             val myRecommendationUrl = requireUrl(url)
```
Comment (RafalOzogJudopay) [inline judokit-android/src/main/java/com/judopay/judokit/android/model/RecommendationConfiguration.kt:50]: Why in line 50 it is okay to have it all in one line, but in "val myRsaKey = requireNotNullOrEmpty(" it has been changed to make new line after '='? I wonder, what is the rule of thumb we should follow to keep it consistent.

File: judokit-android/src/main/java/com/judopay/judokit/android/ui/cardentry/CardEntryViewModel.kt [hunk 5]
```diff
@@ -302,155 +312,178 @@ class CardEntryViewModel(
...
 
-private val Judo.continueButtonText: Int
-    get() = when (paymentWidgetType) {
+private fun Judo.submitButtonText(options: CardEntryOptions): Int =
+    when (paymentWidgetType) {
+        PaymentWidgetType.REGISTER_CARD -> R.string.register_card
         PaymentWidgetType.CREATE_CARD_TOKEN -> R.string.save_card
-        PaymentWidgetType.REGISTER_CARD,
-        PaymentWidgetType.CHECK_CARD,
+        PaymentWidgetType.CHECK_CARD -> R.string.check_card
```
Comment (RafalOzogJudopay) [inline judokit-android/src/main/java/com/judopay/judokit/android/ui/cardentry/CardEntryViewModel.kt:412]: comma and new line here? ;)

File: judokit-android/src/main/java/com/judopay/judokit/android/ui/cardentry/CardEntryViewModel.kt [hunk 6]
```diff
@@ -302,155 +312,178 @@ class CardEntryViewModel(
...
         PaymentWidgetType.CARD_PAYMENT,
         PaymentWidgetType.PRE_AUTH,
         PaymentWidgetType.TOKEN_PAYMENT,
-        PaymentWidgetType.TOKEN_PRE_AUTH -> R.string.continue_text
+        PaymentWidgetType.TOKEN_PRE_AUTH,
+        ->
+            if (uiConfiguration.shouldPaymentButtonDisplayAmount) {
```
Comment (RafalOzogJudopay) [inline judokit-android/src/main/java/com/judopay/judokit/android/ui/cardentry/CardEntryViewModel.kt:418]: I assume this is intended change for strings logic, isnt?

**Review by RafalOzogJudopay — APPROVED** [review]
A huge PR. It took me a while to understand new structure of build-logic, convention plugins, and new configuration logic - but I think I grasp it now.

I've noticed bunch of new things I didn't spot before in our codebase, like 'check' function or R namespace aliases - so good to know, I'll be using them from now on.

I am a little confused about syntax rules you introduce (e.g. comma at the last listed item, spacing with single-line vs multi-line functions). Maybe we could have a call after weekend about this, to clarify these rules, so it would be easier to follow in the future.

Left few comments, please have a look into them - but I am approving this PR now (as these are not stoppers, and I know we want to produce the release asap).

File: judokit-android/api/judokit-android.api (hunk 2 above)
Comment (stefan-tudor) [reply judokit-android/api/judokit-android.api:4989]: https://github.com/Kotlin/binary-compatibility-validator

File: judokit-android/src/main/java/com/judopay/judokit/android/api/AppMetaDataProvider.kt (hunk 3 above)
Comment (stefan-tudor) [reply judokit-android/src/main/java/com/judopay/judokit/android/api/AppMetaDataProvider.kt:26]: `printStackTrace` is forbidden in the project by lint rules - hence, it was removed, but the rest of the code remains unchanged.

File: build-logic/README.md (hunk 1 above)
Comment (stefan-tudor) [reply build-logic/README.md:30]: I will enumerate convenience plugins later

**Pull request merged**